src/transaction/log_replication_atomic.cpp

```diff
@@ -158,6 +161,7 @@ namespace cublog
 
 	    if (m_locked_classes.count (header.trid) > 0)
 	      {
+		discard_caches_for_ddl (thread_entry, header.trid);
```

Comment:
Honest question.
In case of an ATS transaction abort, the ATS will add undo log records for every "redo" log record added by the transaction up to that point.
Thus, after the transaction rollback, the database schema remains unchanged - from a logical point of view.
Is it the same from a physical point of view as well?
Or do some of the OIDs for the class definitions change following the application of the undo log records?
Because, if anything changes, then it makes sense to decache in the same way is done for commit.

Thus, does it make sense to discard the caches at this point?

Thank you.

Reply:
@joohok Did you say that a cache (or some caches?) is updated in a redo function while replicating?

Reply:
And is there no chance a cache is updated while replicating? like
1. an entry for a class is de-cached due to a policy like LRU.
2. a new entry for the class is cached based on the changed physical image of the class while replicating.

If this is true, we should drop the cache after aborting. Isn't it possible?

Reply:
@wrlawodms  
Yes, only the cache related to HFID (heap_Hfid_table) is updated during replication. 

1. I haven't considered it from an LRU perspective, but it seems that de-caching should not be a concern. The main goal is to prevent incomplete information from being cached during the replication process. 
Additionally, there should be some caching during the replication process for de-caching by LRU to take place.

2. Cache is modified only by the logical modifications. The sequence involves 1) a logical change request, 2) utilizing, modifying, and storing cache information, 3) followed by a physical image change. However, since the replicator only modifies the physical image, there are no cache changes except for HFID caches.

Reply:
1. Is not the hfid_cache a target to decache on commit or abort by the replicator?
2. Would it be safer to add assert()s in cache-update functions that it must not be a replicator transaction or thread? Can it be identified?

Reply:
1. yes, because it is already de-cached by replicator in (vacuum_rv_notify_dropped_file ())
---------------------------
BUT .. I should also de-cache the hfid_table here, because I didn't consider the TRUNCATE case. It will have same OID but different HFID, so I have to discard the cache. If it is not de-cached, read transaction can refer to old heap file (heap file is changed by TRUNCATE TABLE statement).   
2. maybe we can identify by the thread type (=TT_REPLICATION_PTS), and I will add an assert for that

Reply:
> an entry for a class is de-cached due to a policy like LRU.

Where is this mechanism implemented? You mean `heap_classrepr_cache::LRU_list` ?

> a new entry for the class is cached based on the changed physical image of the class while replicating.

That for sure will lead to inconsistent cache - the caching logic might even not be able to conclude the caching because heap pages that contain the class information are still scrambled.

This can be mitigated by putting the DDL replication log record sequences under the umbrella of "atomic sequences" (thus the affected heap pages will be kept locked for writing until the image is consistent). And the caching mechanism will not be queued to access these pages to reconstruct the cache.

Reply:
> Where is this mechanism implemented? You mean heap_classrepr_cache::LRU_list ?

Oh, no. LRU is a hypothetical example. If this kind of policy exists, the cache could be inconsistent. So, I just asked.